### PR TITLE
Create Land Obstacle

### LINKED ISSUE #238: Ensure we don't hit land
We are likely starting out a near shore test from Royal Vancouver Yacht Club

https://www.google.com/maps/place/Royal+Vancouver+Yacht+Club/@49.2781497,-123.2026223,14.3z/data=!4m5!3m4!1s0x5486725781a1afe7:0x6f051c87aa6f39c4!8m2!3d49.2727529!4d-123.187769

Sailbot_Shore_Test

```
roslaunch local_pathfinding all.launch AIS_token:=<token> start_lat:=49.276988 start_lon:=-123.187338 goal_lat:=49.325087 goal_lon:=-123.611634 
```

With upwind
```
roslaunch local_pathfinding all.launch AIS_token:=<token> start_lat:=49.276988 start_lon:=-123.187338 goal_lat:=49.325087 goal_lon:=-123.611634 global_wind_direction_degrees:=-20
```

We should ace this test and ensure we don't hit anything

image

Currently, we may go towards land if the wind and goal positions are unperfectly aligned. Likely add an obstacle that represents land.

### PR DESCRIPTION
**Pull request by patrick-5546**
Resolves #238

- Created the `GeneralPolygon` obstacle in the visualizer to represent land
- Created the `land_mass_creator.py` script to generate land mass input data (see discussion below)

#### Additional Dependencies
- Shapely
    ```
    pip install shapely
    ```

#### Usage
The rosparam `land_mass_file` is a path (relative to this repo's root directory) to a land mass file containing its bounding latlons.

```sh
roslaunch local_pathfinding all.launch land_mass_file:=land_masses/land_mass_2.csv
```

image

### CODE REVIEW

**Comment by tylerlum** [comment]
image

Started debugging the issue. Nothing intelligent, just tons of prints with help of vim
image

Stops at
```
    state = sailbot.getCurrentState()
```

Further analysis of prints leads to finding the loop


image
image

Essentially, goal waypoint is always invalid. Must be something wrong with the new polygon

**Comment by tylerlum** [comment]
Played around with it a bit more and got it working! Will let you enjoy the fun of getting it working yourself ;)

**Comment by tylerlum** [comment]
Great work, very clean code additions

**Comment by tylerlum** [comment]
Because the above is a potentially infinite while loop, we should add rospy logs.

**Comment by patrick-5546** [comment]
`land_mass_creator.py` usage:
```
./land_masses/land_mass_creator.py -h

usage: land_mass_creator.py [-h] [-d THRESHOLD_DEPTH] south north west east

Returns locations of shallow water or land, given coordinates

positional arguments:
  south                 bottom latitude bound
  north                 top latitude bound
  west                  left longitude bound
  east                  right longitude bound

optional arguments:
  -h, --help            show this help message and exit
  -d THRESHOLD_DEPTH, --threshold_depth THRESHOLD_DEPTH
                        maximum depth below sea level of the returned location
                        (in metres, exclusive, default 5)
```
```
./land_masses/land_mass_creator.py 49.115757 49.303264 -123.394709 -123.159747

netCDF dimensions: 48.0020833333 49.4979166667 -124.997916667 -123.002083333
nc index bounds: lat=[267,312) lon=[384,441)
Saved to land_masses/land_depth_5m_10_05_2021-00_02_47.(txt|kml)
Took 0.420228004456 seconds
```

**Comment by patrick-5546** [comment]
### Vancouver Land Mass
### KML
image
### TXT > visualizer
image

**Review by tylerlum — APPROVED** [review]
Impressive results and testing. Great work. Let's get this merged in :)

File: python/obstacles.py [hunk 1]
```diff
@@ -51,6 +57,16 @@ def getObstacles(state, referenceLatlon):
     elif obstacle_type == "hybrid_circle":
         for ship in ships:
             obstacles.append(HybridCircleObstacle(ship, position, referenceLatlon))
+
+    # create a land mass obstacle if land_latlons is not empty
+    # path relative to local-pathfinding directory
+    land_mass_file = rospy.get_param('land_mass_file', default='')
+    land_mass_file_abs_path = os.path.join(LOCAL_DIR_ABS_PATH, land_mass_file)
+    if os.path.exists(land_mass_file_abs_path):
```
Comment (tylerlum) [inline python/obstacles.py:65]: Looks like need
```
if os.path.isfile('filename.txt'):
```

**Pull request merged**